enhancements/cluster-logging/cluster-logging-log-forwarding.md

```diff
@@ -99,259 +116,211 @@ This is often required for industries such as the US public sector, healthcare o
-       retry_tag "retry_elasticsearch"
-     </store>
- </match>
-</label>
-<label @SECUREFORWARD_OFFCLUSTER>
-   ...
-</label>
+apiVersion: "logging.openshift.io/v1"
+kind: "ClusterLogForwarder"
+spec:
+  outputs:
+   - name: MyLogs
+     type: Syslog
+     syslog: {Facility: Local0}
+     url: localstore.example.com:9200
+  pipelines:
+   - inputs: [Infrastructure, Application, Audit]
+     outputs: [MyLogs]
 ```
 
-Following is a diagram which identifies the workflow of a message through fluentd as captured by the above configuration.
+#### As a cluster administrator, I want to clearly separate where the logging stack forwards infrastructure and/or audit related logs.
 
-![https://docs.google.com/drawings/d/18FD4c1pBz4e8eJIH37yUwM40ew69g8h1cqPEfOLsuyY/edit?usp=sharing](images/log-forwarding-message-flow.png)
+```
+apiVersion: "logging.openshift.io/v1"
+kind: "ClusterLogForwarder"
+metadata:
+  outputs:
+   - { name: MyInfra, ... }
+   - { name: MyApp, ... }
+   - { name: MyAudit, ... }
+  pipelines:
+   - inputs: [Infrastructure]
+     outputs: [ MyInfra ]
+   - inputs: [Application]
+     outputs: [ MyApp ]
+   - inputs: [Audit]
+     outputs: [ MyAudit ]
+```
 
-### Cluster Logging Operator
-* The operator generates output configuration for the collector
-* The operator specifies in the collector daemonset the secrets needed to communicate to each endpoint:
+### Test Plan
 
-   `/var/run/ocp-collector/secrets/<secretName>/{tls.crt,tls.key,ca-bundle.crt}`
+#### Regression testing
+Translate all existing TP tests to new API, translation should be simple, tests should pass.
 
-    where:
+n#### Unit testing
+* Log forwarding will add unit tests to provide adequate coverage for all changes
+* BDD unit testing will be added to unit testing to make tests goals more expliit,  readable, and obvious
+* Use `go test -cover` and related tools to measure coverage https://blog.golang.org/cover
 
-    * `secretName` is the name of a secret in the `openshift-logging` namespace.
-    * `tls.crt`, `tls.key`, `ca-bundle.crt` are files in the secret
+#### Functional testing
+Go tests that run (in sub-processes or goroutines):
+- a collector instance
+- a dummy log receiver
+- a simulated container generating logs
 
+Verify configurations:
+* Pipelines with multiple, overlapping outputs.
+* TLS server and client authentication.
+* Fluent and syslog protocols.
+* Verify reconnect
+* Other error scenarios.
 
-#### Collector
-* The collectors will be modified to be remove endpoint config specific logic from the start script ; configuration is assumed to be correct and used as provided by the `cluster-logging-operator`
-* Extract all configuration into the collector configuration.
-* Extract the `run.sh` script from the collector image and mount into the deployed pod
+Note: by driving functional tests from `go test` we can get coverage stats integrated with the unit tests for free.
 
-### Test Plan
```

Comment:
A general comment on the test plan: support for Graylog and Splunk are declared as primary goals, but the test plan doesn't address testing against them.

Reply:
Thanks, that is a useful observation. Mr. Sargay Yadriko

Reply:
Any time, Mr. Elen Conwe